frontend/test/resolution/testReturnTypes.cpp

```diff
@@ -1094,6 +1094,172 @@ static void testSelectParams() {
   }
 }
 
+// Assumes loop body will return param string "asdf"
+static void paramLoopTestHelper(Context* context, const char* loopBody,
+                                bool returnFromLoop, bool useEmptyLoop = false) {
+  context->advanceToNextRevision(false);
+  ErrorGuard guard(context);
+
+  // Construct test program
+  std::ostringstream oss;
+  oss << ops;
+  oss << "proc foo() param {\n";
+  oss << "  for param i in " << (useEmptyLoop ? "1..0" : "0..2") << " {\n";
+  oss << loopBody;
+  oss << "  }\n";
+  oss << "  return true;\n";
+  oss << "}\n";
+  oss << "param x = foo();\n";
+  std::string program = oss.str();
+
+  std::cout << "Param loop test program:\n";
+  std::cout << program.c_str() << "\n";
+
+  QualifiedType qt = resolveTypeOfXInit(context,
+                                       program);
+  assert(!qt.isUnknownOrErroneous());
+  if (returnFromLoop) {
+    ensureParamString(qt, "asdf");
+  } else {
+    ensureParamBool(qt, true);
+  }
+
+  assert(guard.realizeErrors() == 0);
+
+  std::cout << "success\n";
+}
+
+// Test returns from within param for loops
+static void testParamLoop() {
+  Context ctx;
+  Context* context = &ctx;
+
+  // Basic case
+  paramLoopTestHelper(context,
+                      R"""(
+                      return "asdf";
+                      )""",
+                      true);
+
+  // Different returns in subsequent iterations
+  paramLoopTestHelper(context,
+                      R"""(
+                      if i == 1 {
+                        return true;
+                      } else {
+                        return "asdf";
+                      }
+                      )""",
+                      true);
+
+  // Return in param TRUE conditional inside param loop iteration
+  paramLoopTestHelper(context,
+                      R"""(
+                      if i == 1 {
+                        return "asdf";
+                      }
+                      )""",
+                      true);
+
+  // Return in param FALSE conditional inside param loop iteration
```

Comment:
I first read this as saying that the return will occur in a param false conditional, and was confused what you mean. Consider wording this as "return statement in", which makes it more clear you're talking about the syntax of the program and not its semantics

Reply:
Sounds good, I think I'll actually apply "return -> return statement" to all the test case descriptions for that reason.